src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.cs

```diff
@@ -298,142 +333,104 @@ private void ProcessEvent(WorkspaceChangeEventArgs args, string eventName)
-                    case WorkspaceChangeKind.AnalyzerConfigDocumentRemoved:
-                    case WorkspaceChangeKind.AnalyzerConfigDocumentChanged:
-                    case WorkspaceChangeKind.AnalyzerConfigDocumentReloaded:
-                        // If an additional file or .editorconfig has changed we need to reanalyze the entire project.
-                        Contract.ThrowIfNull(e.ProjectId);
-                        EnqueueEvent(e.NewSolution, e.ProjectId, InvocationReasons.AdditionalDocumentChanged, eventName);
-                        break;
-
-                    default:
-                        throw ExceptionUtilities.UnexpectedValue(e.Kind);
-                }
-            }
-
-            private void ProcessProjectEvent(WorkspaceChangeEventArgs e, string eventName)
-            {
-                switch (e.Kind)
-                {
-                    case WorkspaceChangeKind.ProjectAdded:
-                        Contract.ThrowIfNull(e.ProjectId);
-                        EnqueueEvent(e.NewSolution, e.ProjectId, InvocationReasons.DocumentAdded, eventName);
-                        break;
-
-                    case WorkspaceChangeKind.ProjectRemoved:
-                        Contract.ThrowIfNull(e.ProjectId);
-                        EnqueueEvent(e.OldSolution, e.ProjectId, InvocationReasons.DocumentRemoved, eventName);
-                        break;
-
-                    case WorkspaceChangeKind.ProjectChanged:
-                    case WorkspaceChangeKind.ProjectReloaded:
-                        Contract.ThrowIfNull(e.ProjectId);
-                        EnqueueEvent(e.OldSolution, e.NewSolution, e.ProjectId, eventName);
-                        break;
-
-                    default:
-                        throw ExceptionUtilities.UnexpectedValue(e.Kind);
-                }
-            }
-
-            private void ProcessSolutionEvent(WorkspaceChangeEventArgs e, string eventName)
-            {
-                switch (e.Kind)
-                {
-                    case WorkspaceChangeKind.SolutionAdded:
-                        EnqueueEvent(e.NewSolution, InvocationReasons.DocumentAdded, eventName);
-                        break;
-
-                    case WorkspaceChangeKind.SolutionRemoved:
-                        EnqueueEvent(e.OldSolution, InvocationReasons.SolutionRemoved, eventName);
-                        break;
+                _eventProcessingQueue.ScheduleTask(
+                    eventName,
+                    async () =>
+                    {
+                        var solutionChanges = newSolution.GetChanges(oldSolution);
 
-                    case WorkspaceChangeKind.SolutionCleared:
-                        EnqueueEvent(e.OldSolution, InvocationReasons.DocumentRemoved, eventName);
-                        break;
+                        // TODO: Async version for GetXXX methods?
+                        foreach (var addedProject in solutionChanges.GetAddedProjects())
```

Comment:
This TODO isn't new, but the original intent behind it seems odd. It looks like we're queueing a work item to diff the projects to then add stuff back to the queue, but there's not really a need to do that asynchronously since GetAddedProjects is going to be relatively fast -- it's just computing the set difference of the IDs and that's that. It's entirely CPU bound with no locking or contention...